lib/ast.js

```diff
@@ -1,66 +1,81 @@
-          declaredFunctions[name] = loc;
-        }
-      } else if (node.type === 'VariableDeclaration') {
-        node.declarations.forEach((decl) => {
-          if (decl.init && decl.init.type === 'FunctionExpression') {
-            if (decl.init.id && decl.init.id.name) {
-              name = decl.init.id.name;
-            } else {
-              name = decl.id.name;
-            }
-            loc = decl.init.body.loc.start;
-            if (vulnerableFunctionNames.includes(name)) {
-              declaredFunctions[name] = loc;
-            }
-          }
-        });
-      } else if (node.type === 'ExportNamedDeclaration') {
-        if (node.declaration.type === 'FunctionDeclaration') {
-          name = node.declaration.id.name;
-          loc = node.declaration.body.loc.start;
-          if (vulnerableFunctionNames.includes(name)) {
-            declaredFunctions[name] = loc;
-          }
+  const parsedScript = acorn.parse(scriptContent,
+    {locations: true, sourceType: 'module'});
+  const body = parsedScript.body;
+  body.forEach(function (node) {
+    inspectNode(node,
+      [],
+      (name, loc) => {
+        const mangled = name.join('.');
+        if (vulnerableFunctionNames.includes(mangled)) {
+          declaredFunctions[mangled] = loc;
         }
+      });
+  });
+
+  return declaredFunctions;
+}
+
+function inspectNode(node, path, cb) {
+  if (!node) {
+    return;
+  }
+  if (node.type === 'FunctionDeclaration') {
+    const loc = node.body.loc.start;
+    const name = node.id.name;
+    cb(path.concat(name), loc);
+  } else if (node.type === 'FunctionExpression') {
+    cb(path, node.body.loc.start);
+  } else if ((node.type === 'ExpressionStatement' && node.expression.right)) {
+    const name = [];
+    const left = node.expression.left;
+
+    if (!left.object) {
+      return;
+    }
+
+    if (left.object.object && left.object.object.name) {
+      name.push(left.object.object.name);
+      if (left.object.property && left.object.property.name) {
+        name.push(left.object.property.name);
+      }
+    } else {
+      name.push(left.object.name);
+    }
+    name.push(left.property.name);
+
+    inspectNode(node.expression.right, path.concat(name), cb);
+  } else if (node.type === 'VariableDeclaration') {
+    node.declarations.forEach((decl) => {
+      let newPath;
+      if (decl.init && decl.init.id && decl.init.id.name) {
+        newPath = path.concat(decl.init.id.name);
+      } else if (decl.id && decl.id.name) {
+        newPath = path.concat(decl.id.name);
       }
+
+      inspectNode(decl.init, newPath, cb);
     });
-  } catch (error) {
-    debug(`Error finding vulnerable methods ${vulnerableFunctionNames}` +
-      ` in script path ${scriptPath}: ${error}`);
+  } else if (node.type === 'ExportNamedDeclaration') {
+    inspectNode(node.declaration, path, cb);
+  } else if (node.type === 'AssignmentExpression') {
+    inspectNode(node.left, path, cb);
```

Comment:
Is there a reason to inspect the left node in this case?

Reply:
I can't think of any *reasonable* reason to do it, but we have other outstanding issues where we're missing functions defined in weird places, so I thought it was worth just recrusing into anything that looked like a normal node.

I can think of syntaxes where you end up with useful code on the "left" of an assignment, but don't know what they parse as right now, or whether anyone would consider them reasonable:

```
> const { a: b = function() { return "hi"; } } = {};
> b()
'hi'
```